samples/nrf9160/nrf_cloud_mqtt_multi_service/src/connection.c

```diff
@@ -0,0 +1,677 @@
+}
+
+/**
+ * @brief Notify that cloud connection is ready
+ */
+static void notify_cloud_ready(void)
+{
+	k_event_post(&cloud_connection_events, CLOUD_READY);
+}
+
+/**
+ * @brief Clear nRF Cloud connection events
+ */
+static void clear_cloud_connection_events(void)
+{
+	k_event_set(&cloud_connection_events, 0);
+}
+
+/**
+ * @brief Await a connection to nRF Cloud (ignoring LTE connection state and cloud readiness)
+ *
+ * @param timeout - The time to wait before timing out
+ * @return true if occurred
+ * @return false if timed out
+ */
+
+static bool await_cloud_connected(k_timeout_t timeout)
+{
+	LOG_DBG("Awaiting Cloud Connection");
+	return k_event_wait(&cloud_connection_events, CLOUD_CONNECTED, false, timeout) != 0;
+}
+
+
+/**
+ * @brief Wait for nRF Cloud readiness
+ *
+ * @param timeout - The time to wait before timing out
+ * @param timeout_on_disconnection - Should cloud disconnection events count as a timeout?
+ * @return true if occurred
+ * @return false if timed out
+ */
+static bool await_cloud_ready(k_timeout_t timeout, bool timeout_on_disconnection)
+{
+	LOG_DBG("Awaiting Cloud Ready");
+	int await_condition = CLOUD_READY;
+	if (timeout_on_disconnection) {
+		await_condition |= CLOUD_DISCONNECTED;
+	}
+	return k_event_wait(&cloud_connection_events, await_condition,
+			    false, timeout) == CLOUD_READY;
+}
+
+bool await_connection(k_timeout_t timeout)
+{
+	return await_lte_connection(timeout) && await_cloud_ready(timeout, false);
+}
+
+bool cloud_is_connected(void)
+{
+	return k_event_wait(&cloud_connection_events, CLOUD_CONNECTED, false, K_NO_WAIT) != 0;
+}
+
+void disconnect_cloud(void)
+{
+	k_event_post(&cloud_connection_events, CLOUD_DISCONNECTED);
+}
+
+bool await_cloud_disconnection(k_timeout_t timeout)
+{
+	return k_event_wait(&cloud_connection_events, CLOUD_DISCONNECTED, false, timeout) != 0;
+}
+
+bool cloud_is_disconnecting(void)
+{
+	return k_event_wait(&cloud_connection_events, CLOUD_DISCONNECTED, false, K_NO_WAIT) != 0;
+}
+
+/**
+ * @brief Handler for date_time events. Used exclusively to detect when we have obtained
+ * a valid modem time.
+ *
+ * @param evt
+ */
+static void date_time_evt_handler(const struct date_time_evt *evt)
+{
+	if (date_time_is_valid()) {
+		notify_date_time_known();
+	}
+}
+
+/**
+ * @brief Handler for events from nRF Cloud Lib.
+ *
+ * @param nrf_cloud_evt Passed in event.
+ */
+static void cloud_event_handler(const struct nrf_cloud_evt *nrf_cloud_evt)
+{
+	switch (nrf_cloud_evt->type) {
+	case NRF_CLOUD_EVT_TRANSPORT_CONNECTED:
+		LOG_DBG("NRF_CLOUD_EVT_TRANSPORT_CONNECTED");
+		/* Notify that we have connected to the nRF Cloud */
+		notify_cloud_connected();
+		break;
+	case NRF_CLOUD_EVT_TRANSPORT_CONNECTING:
+		LOG_DBG("NRF_CLOUD_EVT_TRANSPORT_CONNECTING");
+		break;
+	case NRF_CLOUD_EVT_USER_ASSOCIATION_REQUEST:
+		LOG_DBG("NRF_CLOUD_EVT_USER_ASSOCIATION_REQUEST");
+		/* This is a pre-ready, post-connection event that indicates that nRF Cloud is
```

Comment:
this event means that the user needs to add the device to their nrf cloud account.
see the lte_ble_gateway sample for an example of handling it.

Reply:
Huh -- the documentation here:
https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/libraries/networking/nrf_cloud.html

Suggests that the correct response to `NRF_CLOUD_EVT_USER_ASSOCIATION_REQUEST` is to disconnect and reconnect.

I'll look at the lte_ble_gateway to see what they are doing, but does the documentation there need to be updated, or am I misunderstanding?

Reply:
either way, perhaps the documentation  needs updated to be more clear?

Reply:
It looks like lte_ble_gateway is responding to the event in the same way that this sample responds to the event (by disconnecting from nrf_cloud):

```c

	if (atomic_get(&association_requested)) {
		atomic_set(&association_requested, 0);

		/* after successful association, the device must
		 * reconnect to aws.
		 */
		LOG_INF("Device associated with cloud");
		LOG_INF("Reconnecting for settings to take effect");
		LOG_INF("Disconnecting from cloud...");

		err = nrf_cloud_disconnect();

		if (err == 0) {
			return;
		} else {
			LOG_ERR("Disconnect failed, rebooting...");
			nrf_cloud_error_handler(err);
		}
	}
}
```
This is from `on_cloud_evt_user_associated` in https://github.com/nrfconnect/sdk-nrf/blob/main/samples/nrf9160/lte_ble_gateway/src/main.c

I agree that the documentation could be made clearer, for sure